### DISCUSSION TITLE
other_proc consumes 57% of memory on the node

### DISCUSSION THREAD

**Discussion by boooch** [rabbitmq/rabbitmq-server#5093] [Other] [answered]
After start this metric always grow up to 20-30 Gb in one day. Until Rabbit became unusable.
![image](https://user-images.githubusercontent.com/1180031/175022843-167e9f48-581e-434e-aac6-b50fe63fbb45.png)

I see this other_proc  eating the most amount of memory:
![image](https://user-images.githubusercontent.com/1180031/175022127-4203b26e-b1aa-4485-b5aa-e1f9f98d22b4.png)
This is result of rabbitmq-diagnostics observer
![image](https://user-images.githubusercontent.com/1180031/175023138-06853715-d1b6-4d37-ae1a-ba1901d80df6.png)

It was on Rabbit MQ 3.9.X and Erlang 24 but after updating to RabbitMQ 3.10.5 and Erlang 25.0.2 still the same.

RabbitMQ 3.10.5 
Erlang 25.0.2
Oracle Linux Server 8.5
Server HP ProLiant DL160 Gen9 
Memory 128Gb

**Comment by michaelklishin** [comment]
There is no evidence of a memory leak. A `gen_event` process has a massive backlog of [Erlang] messages it is yet to process.

Most likely this is a result of a few factors or some combination of them:

 * The node is underpowered in terms of available CPU cores
 * There is a high churn of connections, channels, queues or bindings going on
 * [Internal event consumption](https://rabbitmq.com/logging.html#internal-events) is used in this cluster and the consumer does not keep up

Increasing the number of CPU cores available to the node and reducing the churn applications produce
should help, as might disabling the event exchange plugin.

**Comment by boooch** [comment]
1  No problems with CPU
![image](https://user-images.githubusercontent.com/1180031/175035805-40ab41b8-b24d-4262-8481-be1c802318b1.png)
2 If this is high ?
![image](https://user-images.githubusercontent.com/1180031/175036615-e178ff4a-c435-4d05-83b9-79ad6d932fb7.png)
3 Please can you comment how to check this?

By the way, I have server with less powerfull hardware and with the same load but RabbitMQ 3.6.15 and Erlang R16B03-1  and this Rabbit have uptime 312 days and spend only 1 Gb and no freezes at all.
But new one awais freezes on readings and spend lot of memory (uptime 3+ hours)
![image](https://user-images.githubusercontent.com/1180031/175038162-5f0c690b-ed7f-4c71-955d-088ed8158909.png)

**Reply by mkuratczyk** [reply]
Could you try to simulate your workload using https://perftest.rabbitmq.com/? That would allow us to run the tests and hopefully optimize some stuff.

**Comment by michaelklishin** [comment]
It's not the number of entities (connections or queues) that's important here. It's the number of internally emitted events, which is
heavily dependent on [connection and queue churn](https://rabbitmq.com/connections.html).

RabbitMQ 3.6 had few types of events, so high churn probably did not affect it as much. For example, it did not have an event for failed authentication attempts. Do that 10K times a second and the `rabbit_event` process can begin accumulating a backlog of events to process.

Logs and churn metrics available via the management UI and Prometheus/Grafana will help narrow things down. I doubt this can be reproduced with PerfTest because it's not the message rates that are important here. Something heavily floods `rabbit_event` with events and/or the node is deprived of CPU resources.

**Comment by boooch** [comment]
Is any method to clean this backlog "on the fly" ? Or may be settings to not generate? Because this is not the thing I want from Rabbit. I only want it to fast receive messages and fast pass on. And memory only for storing messages not for some unknown events.

**Reply by boooch** [reply]
For now only the way I see is to setup old version of RabbitMQ or reboot Rabbit every 2-3 hours :-) But this is very funny sulution.

**Reply by boooch** [reply]
This is my Churn statistics. Channel operations too high?
![image](https://user-images.githubusercontent.com/1180031/175098124-fe9e8ad2-c726-4c44-b0bf-d431625321d8.png)

**Reply by lukebakken** [reply]
Yes, far too many channel operations.

**Comment by boooch** [comment]
May be I do something wrong on client side?
![image](https://user-images.githubusercontent.com/1180031/175099860-2762e09b-5917-40b4-9538-33c8be7dea11.png)

**Comment by lukebakken** [comment, marked as answer]
It looks like you're using PHP. I suggest using this proxy: https://github.com/cloudamqp/amqproxy

**Reply by boooch** [reply]
Thank you! That helped. 

**Before:**
![image](https://user-images.githubusercontent.com/1180031/175279305-c186f4c0-b938-43f4-aef3-9f4a8be0ec27.png)
**After:**
![image](https://user-images.githubusercontent.com/1180031/175279443-0b3e29bc-a559-4a89-9292-77d01d8bc947.png)

**Reply by lukebakken** [reply]
Great! Thank you for letting us know. I'm assuming this fixed the memory use issue as well.

**Reply by michaelklishin** [reply]
Yup, that must have eliminated the connection churn and thus relieved `rabbit_event`. Once the backlog of unprocessed events has been cleared, so the memory pressure has expectedly dropped as well.

**Reply by boooch** [reply]
Yes, memory issue gone:
![image](https://user-images.githubusercontent.com/1180031/175881553-958646fd-1de5-41b7-8c8a-66a749c1e599.png)

But really want to know is the any hope to remove this issue without using additional service as proxy? May be some settings?

**Reply by mkuratczyk** [reply]
You don't need to introduce proxies, if your applications can maintain a long-lived connection. Connection churn is always an anti-pattern, not just for RabbitMQ. Having said that, we are aware of at least one thing that makes this particularly bad right now and we are going to either improve it, or disable the feature that causes this by default (connection tracking for the purpose of per-user/per-vhost limits; currently it's always enabled). Also, the whole backend for how we store this information is going to be replaced in the future (Khepri will replace Mnesia), which will completely change how storing this information is handled.

**Reply by boooch** [reply]
I cannot change this logic (lot of reconnects) because I use it from php via nginx+php-fpm. But using **pconnect** instead of  **connect** allows me to reuse connections, but I have to recreate channels and this was issue.  In older versions of RabbitMQ  this architecture was without any leaks and issues.
